allennlp/nn/beam_search.py

```diff
@@ -25,6 +27,365 @@
+    distribution.
+    Registered as a `Sampler` with name "top-k".
+    """
+
+    def __init__(
+        self,
+        k: int = 1,
+        temperature: float = 1.0,
+        filter_val: float = -float("inf"),
+        with_replacement: bool = False,
+    ):
+        if not isinstance(k, int) or k < 1:
+            raise ValueError("k must be a positive integer")
+        self.k = k
+        self.temperature = temperature or 1.0
+        self.filter_val = min_value_of_dtype(torch.float)
+        self.with_replacement = with_replacement
+
+    @overrides
+    def sample_nodes(
+        self, log_probs: torch.Tensor, per_node_beam_size: int, state: StateType
+    ) -> Tuple[torch.Tensor, torch.Tensor, StateType]:
+        assert self.k <= len(log_probs[-1])
+
+        if self.temperature != 1.0:
+            _log_probs = torch.nn.functional.log_softmax(log_probs / self.temperature, dim=-1)
+        else:
+            _log_probs = log_probs
+
+        # Find the indices that are not to be selected from
+        min_threshold = torch.topk(_log_probs, self.k)[0][..., -1].unsqueeze(dim=-1)
+        filtered_indices = _log_probs < min_threshold
+
+        # Prevent the filtered indiceds from being selected
+        _log_probs[..., filtered_indices] = self.filter_val
+        filtered_probabilities = torch.nn.functional.softmax(_log_probs, dim=-1)
+
+        # Shape: (batch_size*beam_size, per_node_beam_size)
+        selected_indices = torch.multinomial(
+            filtered_probabilities, per_node_beam_size, replacement=self.with_replacement
+        )
+
+        # Return (selected log probabilities, selected classes)
+        # shape: (batch_size*beam_size, num_samples), (batch_size*beam_size, num_samples)
+        return torch.gather(_log_probs, 1, selected_indices), selected_indices, state
+
+
+@Sampler.register("top-p")
+class TopPSampler(Sampler):
+    """
+    Represents a `Sampler` which redistributes the probability mass function among
+    the top choices with a cumulative probability of at least `p` then selects from that subset
+    `p` if minimum cumulative probability of highest-probability options that the returned
+    choice will be selected from `temperature` modules the probabilitis of the selected tokens.
+    A `temperature` below 1.0 produces a sharper probability distribution and a `temperature`
+    above 1.0 produces a flatter probability distribution.
+    Registered as a `Sampler` with name "top-p".
+    """
+
+    def __init__(
+        self,
+        p: float = 0.9,
+        temperature: float = 1.0,
+        filter_val: float = -float("inf"),
+        with_replacement: bool = False,
+    ):
+        if not isinstance(p, float) or p < 0.0 or p > 1.0:
+            raise ValueError("p must be a non-negative float no greater than 1.0")
+        self.p = p
+        self.temperature = temperature or 1.0
+        self.filter_val = min_value_of_dtype(torch.float)
```

Comment:
Same suggestion as above, we should create this value on the fly.

```suggestion
        self.filter_val = min_value_of_dtype(torch.float)
```